### DISCUSSION TITLE
[v4] Will there be a way to opt out of native cascade layers?

### DISCUSSION THREAD

**Discussion by wongjn** [tailwindlabs/tailwindcss#13188] [Ideas]
The problem with CSS cascade layers is if other third-party CSS (that are supposed to be component level) does not use CSS layers, they will take precedence over CSS in any cascade layer.

One could argue that `!important` would solve this, but then one cannot use `!important` to increase specificity of one utility over another, since they all already have `!important`.

**Comment by adamwathan** [comment]
So right now you can technically opt-out of them by importing the layers separately and just not putting them in layers:

```diff
- @import "tailwindcss/preflight" layer(base);
- @import "tailwindcss/utilities" layer(utilities);
+ @import "tailwindcss/preflight";
+ @import "tailwindcss/utilities";
```

...but there will be downsides because we've adjusted the specificity of a few things with the assumption that things would be in layers and wouldn't conflict with each other.

With third-party CSS though you could always put those styles in a layer as well right? Even if it's a CDN thing you are doing at run-time you can do something like:

```html
<html>
<head>
  <style>
    @import "https://cdnjs.cloudflare.com/ajax/libs/pikaday/1.6.0/css/pikaday.min.css" layer(components);
  </style>
</head>
<!-- ... -->
```

There's some talk of supporting importing stylesheets directly into a layer when using `<link>` too but nothing beyond early conversations so far:

https://github.com/w3c/csswg-drafts/issues/5853

**Reply by wongjn** [reply]
Thank you for those points!

For further context, I mainly work with Tailwind in the context of Drupal themes and WordPress block themes, which are PHP content management systems that have their own "base" or "component" CSS for elements/components that would be output via `<link>` or inline `<style>` tags but where there isn't a great API for modifying them to fit within a cascade layer system. Especially with WordPress's Gutenberg editor where CSS is injected at runtime with zero API surface for editing their CSS without totally nuking it for your own "fork" in your own theme that would then be out of date as the core updates. On the other hand, I could also cede that this is not Tailwind's problem either 😅 

Yes, I did research and come across that proposal too before this post. Great minds think a like perhaps?

**Reply by StevenStavrakis** [reply]
In my case, it wouldn't be possible. I'm trying to use TW to style Obsidian plugins, and I have no control over how the app inserts its styles. Even if I did, I would not want to mess around with the app styles anyway.

**Reply by LostKobrakai** [reply]
We're rendering pdfs using prince using tailwindcss and styles shared with a web frontend. There's simply no support for @layer or @property (at least that's what it already complained about) in that engine. Sounds like we'll need to stick with tailwind 3

**Reply by JunHaoShih** [reply]
So basically anyone that has to deal with third party css that you have zero controll are stucked in Tailwind v3? Jesus

**Reply by totti-rdz** [reply]
@adamwathan Do you have an example of things conflicting each others, when opting out of native css layering?

**Comment by jacobg** [comment]
How do you opt out of css layers using @tailwindcss/vite plugin?

**Reply by wongjn** [reply]
Change your CSS like so:

```diff
-@import "tailwindcss";
+@import "tailwindcss/theme.css";
+@import "tailwindcss/preflight.css";
+@import "tailwindcss/utilities.css";
```

**Reply by ADTC** [reply]
Thanks! Since Astro uses Vite under the hood, I was able to use this in an Astro project.
